pkg/ottl/ottlfuncs/func_replace_all_patterns.go

```diff
@@ -59,42 +61,126 @@ func replaceAllPatterns[K any](target ottl.PMapGetter[K], mode string, regexPatt
 		if err != nil {
 			return nil, err
 		}
-		if fn.IsEmpty() {
-			replacementVal, err = replacement.Get(ctx, tCtx)
-			if err != nil {
-				return nil, err
-			}
-		} else {
-			fnVal := fn.Get()
-			replacementExpr, errNew := fnVal.Get(&replaceAllPatternFuncArgs[K]{Input: replacement})
-			if errNew != nil {
-				return nil, errNew
-			}
-			replacementValRaw, errNew := replacementExpr.Eval(ctx, tCtx)
-			if errNew != nil {
-				return nil, errNew
-			}
-			replacementValStr, ok := replacementValRaw.(string)
-			if !ok {
-				return nil, fmt.Errorf("replacement value is not a string")
-			}
-			replacementVal = replacementValStr
+		replacementVal, err = replacement.Get(ctx, tCtx)
+		if err != nil {
+			return nil, err
 		}
 		updated := pcommon.NewMap()
 		updated.EnsureCapacity(val.Len())
 		val.Range(func(key string, originalValue pcommon.Value) bool {
 			switch mode {
 			case modeValue:
 				if compiledPattern.MatchString(originalValue.Str()) {
-					updatedString := compiledPattern.ReplaceAllString(originalValue.Str(), replacementVal)
-					updated.PutStr(key, updatedString)
+					if !fn.IsEmpty() {
+						var updatedString string
+						var groupNum int
+						foundMatch := false
+						updatedString = originalValue.Str()
+						submatches := compiledPattern.FindAllStringSubmatchIndex(updatedString, -1)
+						// check if the string is indeed a group number string
+						if !IsGroupNumString(replacementVal) {
+							updatedString = compiledPattern.ReplaceAllString(originalValue.Str(), replacementVal)
+							updated.PutStr(key, updatedString)
+							return true
+						}
+						// parse the group number string to get actual group number
+						captureGroup := strings.TrimPrefix(replacementVal, "$")
+						groupNum, err = strconv.Atoi(captureGroup)
+						if err != nil {
+							return false
+						}
+						for _, submatch := range submatches {
```

Comment:
Would something like the following remove the need to handle retrieving the submatches ourselves?

```golang
old := compiledPattern.ExpandString([]byte{}, replacementVal, originalVal.Str(), submatches)
new := optionalFn(old) // Pseudocode for the optional transformation function
updatedString = strings.ReplaceAll(updatedString, old, new)
```

Reply:
I tried this approach and I think the limitation here was that it wasn't able to hash multiple matches for the same capture group. It also doesn't know how to exclude full matches, which is a part of `submatches`

Reply:
I think we should be obtaining a full match, the submatches should just be used to populate any references in the replacement string.

As for getting multiple matches, that should be handled by `FindAllStringSubmatchIndex`, right? My understanding is that each item in the returned array contains the full match and all submatches. So looping through each match and calling `ExpandString` will give us the replacement value with the values for the submatches on each capture group for the match.

Reply:
Used `ExpandString` in the latest [commit](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/29408/commits/d80955615abddfd95e3512fb35b4aa8667457d90)